test/unit/puppetlabs/services/jruby/jruby_puppet_service_test.clj

```diff
@@ -289,3 +289,128 @@
+                                                         nil)
+         (is (= "1234prod" (jruby-protocol/get-environment-class-info-tag
+                            service
+                            "production")))
+         (is (nil? (jruby-protocol/get-environment-class-info-tag
+                    service
+                    "test")))
+         (jruby-protocol/set-environment-class-info-tag! service
+                                                         "test"
+                                                         "1234test"
+                                                         nil)
+         (is (= "1234prod" (jruby-protocol/get-environment-class-info-tag
+                            service
+                            "production")))
+         (is (= "1234test" (jruby-protocol/get-environment-class-info-tag
+                            service
+                            "test"))))
+       (let [production-first-updated
+             (jruby-protocol/get-environment-class-info-tag-last-updated
+              service
+              "production")
+             test-first-updated
+             (jruby-protocol/get-environment-class-info-tag-last-updated
+              service
+              "test")]
+         (testing "when environment info reset in the cache"
+           (is (not (nil? production-first-updated)))
+           (is (not (nil? test-first-updated)))
+           (jruby-protocol/set-environment-class-info-tag!
+            service
+            "production"
+            "5678prod"
+            production-first-updated)
+           (is (= "5678prod" (jruby-protocol/get-environment-class-info-tag
+                              service
+                              "production")))
+           (is (= "1234test" (jruby-protocol/get-environment-class-info-tag
+                              service
+                              "test")))
+           (is (= test-first-updated
+                  (jruby-protocol/get-environment-class-info-tag-last-updated
+                   service
+                   "test")))
+           (let [production-second-update
+                 (jruby-protocol/get-environment-class-info-tag-last-updated
+                  service
+                  "production")]
+             (is (not= production-first-updated production-second-update))
+             ;; Set a new tag with a different "last-updated" date than what
+             ;; should be in the cache right now - (dec
+             ;; production-first-updated) - in order to validate that what is
+             ;; already in the cache will just be retained.  This simulates
+             ;; what would happen if the tag had been updated again in
+             ;; between the prior 'get...last-updated' call and the
+             ;; next 'set...tag' call.
+             (jruby-protocol/set-environment-class-info-tag!
+              service
+              "production"
+              "89abprod"
+              (dec production-first-updated))
```

Comment:
No, this was totally contrived.  I'll re-write this bit to do a `mark-expired` call before the second `set-environment-class-info-tag!` call and pass the original `production-first-updated` into that second call.  That's the real case that I'd expect to see in production anyway.